### PR TITLE
ADD: Fix to mlm memory consumption

### PR DESCRIPTION
**Pull request by benleetownsend**

### CODE REVIEW

File: finetune/nn/target_blocks.py [hunk 1]
```diff
@@ -54,8 +54,7 @@ def masked_language_model(*, X, M, mlm_weights, mlm_positions, mlm_ids, embed_we
     hidden = merge_leading_dims(hidden, 3)
     batch, seq, _ = shape_list(X)
     with tf.variable_scope('model/masked-language-model'):
-        gathered_hidden = gather_indexes(hidden, mlm_positions)
-        
+        gathered_hidden = merge_leading_dims(tf.gather(hidden, mlm_positions, batch_dims=1), 2)
```
Comment (lhz1029) [inline finetune/nn/target_blocks.py]: nice

File: finetune/target_models/masked_language_model.py [hunk 2]
```diff
@@ -29,9 +29,9 @@ def feed_shape_type_def(self):
                 {
                     "tokens": TS([None, 2]),
                     "mask": TS([None]),
-                    "mlm_weights": TS([self.config.max_masked_tokens * self.config.batch_size]),
```
Comment (benleetownsend) [inline finetune/target_models/masked_language_model.py:32]: This is the key original issue. We previously were padding out to [Batch, max_masked * batch] tokens

File: finetune/nn/target_blocks.py [hunk 3]
```diff
@@ -84,13 +83,12 @@ def masked_language_model(*, X, M, mlm_weights, mlm_positions, mlm_ids, embed_we
         mlm_ids = tf.reshape(mlm_ids, [-1])
         mlm_weights = tf.reshape(mlm_weights, [-1])
 
-        log_probs = tf.nn.log_softmax(logits, axis=-1)
-        one_hot_labels = tf.one_hot(mlm_ids, depth=n_vocab, dtype=tf.float32)
-        per_example_loss = -tf.reduce_sum(log_probs * one_hot_labels, axis=[-1])
```
Comment (benleetownsend) [inline finetune/nn/target_blocks.py:89]: this mul by one_hot is pretty memory hungry but maybe a little bit faster. Need to run some more benchmarks.

File: finetune/nn/target_blocks.py [hunk 4]
```diff
@@ -84,13 +83,12 @@ def masked_language_model(*, X, M, mlm_weights, mlm_positions, mlm_ids, embed_we
...
-        numerator = tf.reduce_sum(mlm_weights * per_example_loss)
-        denominator = tf.reduce_sum(mlm_weights) + 1e-5
-        mlm_loss = numerator / denominator
-
+        mlm_loss = tf.contrib.losses.sparse_softmax_cross_entropy(
```
Comment (lhz1029) [inline finetune/nn/target_blocks.py:73]: :+1:

File: finetune/target_models/masked_language_model.py [hunk 5]
```diff
@@ -29,9 +29,9 @@ def feed_shape_type_def(self):
...
-                    "mlm_ids": TS([self.config.max_masked_tokens * self.config.batch_size]),
-                    "mlm_positions": TS([self.config.max_masked_tokens * self.config.batch_size]),
+                    "mlm_weights": TS([None]),
```
Comment (lhz1029) [inline finetune/target_models/masked_language_model.py:32]: sweet. one thing to note is that now the len dim could be variable, which means that the return type if predicting in a batch might be an iterable instead of an array. though it aren't really doing any batch prediction anyway so it probably doesn't matter.

Comment (benleetownsend) [reply finetune/target_models/masked_language_model.py:32]: I'm not sure I follow.

Comment (madisonmay) [reply finetune/target_models/masked_language_model.py:32]: @lhz1029 that's not the case. the batch dim is flattened in current predict anyhow.

Comment (lhz1029) [reply finetune/target_models/masked_language_model.py:32]: it's not the case that the seq len dim could be variable?

File: finetune/nn/target_blocks.py (hunk 3 above)
Comment (lhz1029) [reply finetune/nn/target_blocks.py:89]: What was the output of these benchmarks?

File: finetune/target_models/masked_language_model.py [hunk 6]
```diff
@@ -88,6 +92,9 @@ class MaskedLanguageModel(BaseModel):
     :param config: A :py:class:`finetune.config.Settings` object or None (for default config).
     :param \**kwargs: key-value pairs of config items to override.
     """
+
+    defaults = dict(low_memory_mode=True, batch_size=48, xla=True)
```
Comment (lhz1029) [inline finetune/target_models/masked_language_model.py:96]: oh so we did see speed ups with xla = True?

**Review by lhz1029 — APPROVED** [review]

File: finetune/target_models/masked_language_model.py (hunk 5 above)
Comment (lhz1029) [reply finetune/target_models/masked_language_model.py:32]: Oh whoops I was thinking about an imaginary `predict` method in MLM that does not currently exist

File: finetune/target_models/masked_language_model.py (hunk 6 above)
Comment (benleetownsend) [reply finetune/target_models/masked_language_model.py:96]: Yeah, there is a non-negligible overhead for compilation though, so maybe it doesn't make sense for short runs/debugging.

**Review by madisonmay — APPROVED** [review]
